File: ant_plus_sensor/utils/average.js
/**
 * Calculate a rolling average on a set of data points
 */
class RollingAverage {
  /**
   * Create a rolling average counter
   *
   * @param {number} interval Length of rolling interval window in milliseconds
   */
  constructor(interval) {
    this.interval = interval;
    this.points = [];
  }

  /**
   * Add data point to average
   *
   * @param {number} value Value
   */
  add(value) {
    const time = Date.now();
    this.points.push([time, value]);
  }

  /**
   * Calculate the current average value over the rolling interval
   *
   * @returns {number} Average value
   */
  average() {
    // Remove all stale values
    const time = Date.now();
    this.points = this.points.filter((x) => x[0] >= time - this.interval);

    const sum = this.points.reduce((acc, curr) => acc + curr[1], 0);
    const average = sum / this.points.length;
    return average;
  }
}

module.exports = RollingAverage;

File: ant_plus_sensor/python_ant_plus_sensor/utils/average.py
import functools
import time as time_mod

"""
Calculate a rolling average on a set of data points
"""
class RollingAverage:
    def __init__(self, interval):
        """

        Create a rolling average counter
        
        @param interval (float): Length of rolling interval window in milliseconds

        """
        self.interval = interval
        self.points = []

    def add(self, value):
        """

        Add data point to average
        
        @param value (number): Value
        
        """
        time = time_mod.time() * 1000
        self.points.append([time, value])
    
    def average(self):
        """

        Remove all stale values

        @return average (float): Average

        """
        time = time_mod.time() * 1000;
        self.points = filter(lambda x: x[0] >= time - self.interval, self.points)
        sum = functools.reduce(lambda acc, curr: acc + curr[1], self.points, 0)
        average = sum / len(self.points)
        return average
